Debian 10/Win8.1 fixes: Allow explicitly using SSLv3, TLS1.0/1.1

krwq opened the pull request:
Fixes: #38444
Possibly fixes some other debian 10 issues (I've seen 50+ failures locally)

This fixes couple of things:
- Allow explicitly using SSLv3, TLS1.0/1.1 - before this fix OpenSSL defaulted to always disable them on some OSes (if you use SslProtocols.None this change has no effect for that)
- Same way lower bar for protocols when using AllowNoEncryption/NoEncryption
- Fixes Win 8.1 issue where explicitly setting protocols to TLS1.2 or higher and SSL2 at the same time is causing negotiation to always fail
- Convert System.Net.Security tests to not require IPv6 (they don't need it and it's disabled on docker by default)
- Couple of test fixes

With this my test run is green on Debian10 docker image

krwq commented on src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c:
```diff
@@ -319,6 +324,12 @@ CryptoNative_SslCtxSetCertVerifyCallback(SSL_CTX* ctx, SslCtxSetCertVerifyCallba
 
+void ResetProtocolRestrictions(SSL_CTX* ctx)
+{
+    SSL_CTX_set_min_proto_version(ctx, 0);
```
I didn't get any local failures without adding this to shim (I did for any other API) - presumably it's not needed?

krwq replied:
CI didn't forgive me though :smile:

bartonjs replied:
If you ran `src/Native/build-native.sh` that'll build non-portable, so the shim definition isn't required.  You can also build it portable (`src/Native/build-native.sh -portable`).   CI builds it both ways (we publish portable, RHEL's packages (and other future distro-provided packages) use non-portable).

bartonjs replied:
These functions are new in OpenSSL 1.1.0, so they'll need to be FALLBACK_FUNCTION defined.  You can put the body of ResetProtocolRestrictions in `#if NEED_OPENSSL_1_0` to avoid needing to define non-portable equivalence.

FYI: Building on Ubuntu 16.04 will give you OpenSSL 1.0.x headers, 18.04 is 1.1.0 headers, and 19.04 is 1.1.1 headers.

krwq replied:
I've ended up expanding those macros since they use stuff which we already have available

krwq commented on src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c:
```diff
@@ -152,9 +152,14 @@ void CryptoNative_SetProtocolOptions(SSL_CTX* ctx, SslProtocols protocols)
     // OpenSSL 1.0 calls this long, OpenSSL 1.1 calls it unsigned long.
 #pragma clang diagnostic push
 #pragma clang diagnostic ignored "-Wsign-conversion"
+
```
remove extra enter

krwq commented:
marking as WIP: need to investigate if Windows failures are by design or a product bug

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c:
```diff
@@ -319,6 +324,12 @@ CryptoNative_SslCtxSetCertVerifyCallback(SSL_CTX* ctx, SslCtxSetCertVerifyCallba
     SSL_CTX_set_cert_verify_callback(ctx, callback, arg);
 }
 
+void ResetProtocolRestrictions(SSL_CTX* ctx)
```
```suggestion
static void ResetProtocolRestrictions(SSL_CTX* ctx)
```

It doesn't need to be accessed outside this file, right?

krwq replied:
I always thought this is no-op in C but I think this will suppress the warning I was seeing

bartonjs replied:
Nah, it just means something completely different than in C#: This function is not callable outside of this file.

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h:
```diff
@@ -398,3 +398,6 @@ Looks up a cipher by the IANA identifier, returns a shared string for the OpenSS
 DLLEXPORT const char* CryptoNative_GetOpenSslCipherSuiteName(SSL* ssl, int32_t cipherSuite, int32_t* isTls12OrLower);
+
+/* Internal APIs */
+void ResetProtocolRestrictions(SSL_CTX* ctx);
```
I don't see any need for this to be in the header.

krwq replied:
actually I'm getting warnings which we treat as errors without it (that's why I added it in the first place). Will see if there is any way to get around that (I could probably move this to the .c file though)

krwq replied:
nvm, just read the comment above

krwq commented:
Win 8.1 failures ended up being a product issue (I added comment in the code and modified first post explaining the issue).

Now need to fix up the OpenSSL shim and it's ready to go

krwq commented:
I think this should be green now and ready for review but will wait for green CI before removing `[WIP] ` prefix

bartonjs commented on src/System.Net.Security/src/System/Net/Security/SecureChannel.cs:
```diff
@@ -199,7 +199,7 @@ private X509Certificate2 EnsurePrivateKey(X509Certificate certificate)
                 // Protecting from X509Certificate2 derived classes.
                 X509Certificate2 certEx = MakeEx(certificate);
-                
+
```
Since there are no technical edits in this file it would be ideal if it was no longer part of this change.

`git checkout dotnet/master -- src/System.Net.Security/src/System/Net/Security/SecureChannel.cs`, IIRC (assuming you have a remote named `dotnet` corresponding to dotnet/corefx)

krwq replied:
will revert, I did manual revert, didn't notice (`-w`) this

bartonjs commented on src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:
```diff
@@ -215,7 +215,7 @@ public static unsafe SecurityStatusPal EncryptMessage(SafeDeleteContext security
             // Copy the input into the output buffer to prepare for SCHANNEL's expectations
             input.Span.CopyTo(new Span<byte>(output, headerSize, input.Length));
-            
+
```
Again, it'd be nice if files with no technical edits weren't part of the PR.

krwq replied:
will revert

bartonjs commented on src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs:
```diff
@@ -52,6 +52,21 @@ internal SslAuthenticationOptions(SslServerAuthenticationOptions sslServerAuthen
             CipherSuitesPolicy = sslServerAuthenticationOptions.CipherSuitesPolicy;
         }
 
+        static private SslProtocols FilterOutIncompatibleSslProtocols(SslProtocols protocols)
```
```suggestion
        private static SslProtocols FilterOutIncompatibleSslProtocols(SslProtocols protocols)
```

(canonical ordering)

bartonjs commented on src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs:
```diff
@@ -36,7 +36,7 @@ internal SslAuthenticationOptions(SslServerAuthenticationOptions sslServerAuthen
             ApplicationProtocols = sslServerAuthenticationOptions.ApplicationProtocols;
             CheckCertName = false;
-            EnabledSslProtocols = sslServerAuthenticationOptions.EnabledSslProtocols;
+            EnabledSslProtocols = FilterOutIncompatibleSslProtocols(sslServerAuthenticationOptions.EnabledSslProtocols);
```
When done at this level, does the user have visibility to this filter having kicked in?

krwq replied:
I don't think so but will double check (I think this would be fine either way)

krwq replied:
https://source.dot.net/#System.Net.Security/System/Net/Security/SslAuthenticationOptions.cs,9b19271fc222aad5,references says we only pass it in

krwq commented:
Mergin with OSX failures since they seem unrelated (most of the recent PRs have some OSX failures, it doesn't allow me to open the logs though) and also same PR before couple of tiny cosmetic changes was green before (d6802a760f8ce9d3379570a56c950f1009741c31 specifically)
